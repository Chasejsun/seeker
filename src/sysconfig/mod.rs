mod net;
mod proc;

pub use net::{setup_ip, DNSSetup};
pub use proc::sys::{list_system_proc_socks, list_user_proc_socks};
