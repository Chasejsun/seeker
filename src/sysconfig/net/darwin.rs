use smoltcp::wire::{IpAddress, IpCidr};
use std::process::Command;
use tracing::info;

pub struct DNSSetup;

impl DNSSetup {
    pub fn new() -> Self {
        info!("setup dns");
        let output = Command::new("networksetup")
            .args(&["-setdnsservers", "Wi-Fi", "127.0.0.1"])
            .output()
            .expect("setup local dns");
        if !output.status.success() {
            panic!(
                "stdout: {}\nstderr: {}",
                std::str::from_utf8(&output.stdout).expect("utf8"),
                std::str::from_utf8(&output.stderr).expect("utf8")
            );
        }
        DNSSetup
    }
}

impl Drop for DNSSetup {
    fn drop(&mut self) {
        info!("clear dns");
        let output = Command::new("networksetup")
            .args(&["-setdnsservers", "Wi-Fi", "empty"])
            .output()
            .expect("clear local dns");
        if !output.status.success() {
            panic!(
                "stdout: {}\nstderr: {}",
                std::str::from_utf8(&output.stdout).expect("utf8"),
                std::str::from_utf8(&output.stderr).expect("utf8")
            );
        }
    }
}

pub fn setup_ip(tun_name: &str, ip: IpAddress, cidr: IpCidr) {
    let ip_s = ip.to_string();
    let output = Command::new("ifconfig")
        .args(&[tun_name, &ip_s, &ip_s])
        .output()
        .expect("run ifconfig");
    if !output.status.success() {
        panic!(
            "stdout: {}\nstderr: {}",
            std::str::from_utf8(&output.stdout).expect("utf8"),
            std::str::from_utf8(&output.stderr).expect("utf8")
        );
    }
    let output = Command::new("route")
        .arg("add")
        .arg(cidr.to_string())
        .arg(ip_s)
        .output()
        .expect("add route");
    if !output.status.success() {
        panic!(
            "stdout: {}\nstderr: {}",
            std::str::from_utf8(&output.stdout).expect("utf8"),
            std::str::from_utf8(&output.stderr).expect("utf8")
        );
    }
}
