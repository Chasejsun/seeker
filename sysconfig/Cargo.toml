[package]
name = "sysconfig"
version = "0.1.0"
authors = ["gfreezy <gfreezy@gmail.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[dependencies]
failure = "0.1.5"
tracing = "0.1"
libc = "0.2.62"

[target.'cfg(target_os="macos")'.dependencies]
libproc = { git = "https://github.com/gfreezy/libproc-rs", rev = "bffc2c4" }
