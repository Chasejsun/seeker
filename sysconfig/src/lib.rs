mod net;
mod proc;

pub use net::{setup_ip, DNSSetup};

#[cfg(target_os = "macos")]
pub use proc::sys::{list_system_proc_socks, list_user_proc_socks};
pub use proc::SocketInfo;
