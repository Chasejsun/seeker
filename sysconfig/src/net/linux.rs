use std::fs::OpenOptions;
use std::io::{Read, Seek, SeekFrom, Write};
use std::process::Command;
use tracing::info;

pub struct DNSSetup {
    original: Vec<u8>,
}

impl DNSSetup {
    pub fn new() -> Self {
        info!("setup dns");
        let mut resolv = OpenOptions::new()
            .read(true)
            .write(true)
            .open("/etc/resolv.conf")
            .unwrap();
        let mut buf = vec![];
        let _ = resolv.read_to_end(&mut buf).unwrap();

        info!(
            "original resolve.conf: {}",
            std::str::from_utf8(&buf).unwrap()
        );
        resolv.set_len(0).unwrap();
        resolv.seek(SeekFrom::Start(0)).unwrap();
        resolv.write_all(b"nameserver 127.0.0.1").unwrap();

        DNSSetup { original: buf }
    }
}

impl Drop for DNSSetup {
    fn drop(&mut self) {
        info!("clear dns");
        let mut resolv = OpenOptions::new()
            .write(true)
            .open("/etc/resolv.conf")
            .unwrap();
        resolv.write_all(&self.original).unwrap();
    }
}

pub fn setup_ip(tun_name: &str, ip: &str, _cidr: &str) {
    let output = Command::new("ip")
        .args(&["addr", "add", ip, "dev", tun_name])
        .output()
        .expect("run ip addr");
    if !output.status.success() {
        panic!(
            "stdout: {}\nstderr: {}",
            std::str::from_utf8(&output.stdout).expect("utf8"),
            std::str::from_utf8(&output.stderr).expect("utf8")
        );
    }
    let output = Command::new("ip")
        .args(&["link", "set", tun_name, "up"])
        .output()
        .expect("run ip addr");
    if !output.status.success() {
        panic!(
            "stdout: {}\nstderr: {}",
            std::str::from_utf8(&output.stdout).expect("utf8"),
            std::str::from_utf8(&output.stderr).expect("utf8")
        );
    }
}
