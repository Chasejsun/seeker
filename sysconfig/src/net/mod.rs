#[cfg(any(target_os = "macos", target_os = "ios"))]
#[path = "darwin.rs"]
pub mod sys;

#[cfg(target_os = "linux")]
#[path = "linux.rs"]
pub mod sys;

pub use sys::{setup_ip, DNSSetup};
