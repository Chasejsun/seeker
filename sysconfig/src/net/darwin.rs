use crate::net::run_cmd;
use std::net::IpAddr;
use tracing::info;

pub struct DNSSetup {
    primary_network: String,
    original_dns: String,
}

impl DNSSetup {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        let route_ret = run_cmd("route", &["-n", "get", "0.0.0.0"]);
        let device = route_ret
            .lines()
            .find(|l| l.contains("interface:"))
            .and_then(|l| l.split_whitespace().last())
            .map(|s| s.trim())
            .expect("get primary device");
        info!("Primary device is {}", device);
        let network_services = run_cmd("networksetup", &["-listallhardwareports"]);
        let mut iter = network_services.lines().peekable();
        let network = loop {
            if let Some(line) = iter.next() {
                if let Some(next_line) = iter.peek() {
                    if next_line.split(":").last().map(|l| l.contains(device)) == Some(true) {
                        if let Some(network) = line.split(":").last().map(|s| s.trim()) {
                            break network;
                        }
                    }
                } else {
                    panic!("No primary network found");
                }
            } else {
                panic!("No primary network found");
            }
        };
        info!("Primary netowrk service is {}", network);
        let original_dns = run_cmd("networksetup", &["-getdnsservers", network])
            .lines()
            .filter_map(|l| l.parse::<IpAddr>().ok())
            .map(|ip| ip.to_string())
            .collect::<Vec<_>>()
            .join(" ");

        info!("Original DNS is {}", &original_dns);
        if !original_dns.is_empty() {
            let _ = run_cmd(
                "networksetup",
                &["-setdnsservers", network, "127.0.0.1", &original_dns],
            );
        } else {
            let _ = run_cmd("networksetup", &["-setdnsservers", network, "127.0.0.1"]);
        }
        DNSSetup {
            primary_network: network.to_string(),
            original_dns,
        }
    }
}

impl Drop for DNSSetup {
    fn drop(&mut self) {
        let dns = if self.original_dns.is_empty() || self.original_dns == "127.0.0.1" {
            "empty"
        } else {
            &self.original_dns
        };
        info!("Restore original DNS: {}", dns);

        let _ = run_cmd(
            "networksetup",
            &["-setdnsservers", &self.primary_network, dns],
        );
    }
}

pub fn setup_ip(tun_name: &str, ip: &str, cidr: &str) {
    let _ = run_cmd("ifconfig", &[tun_name, ip, ip]);
    let _ = run_cmd("route", &["add", cidr, ip]);
}
