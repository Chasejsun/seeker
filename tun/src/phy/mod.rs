use async_std::net::driver::Watcher;
use futures::{AsyncRead, AsyncWrite};
use std::io;
use std::io::{Read, Write};
use std::pin::Pin;
use std::task::{Context, Poll};

mod sys;

pub(crate) struct TunSocket {
    watcher: Watcher<sys::TunSocket>,
}

impl TunSocket {
    pub fn new(name: &str) -> TunSocket {
        TunSocket {
            watcher: Watcher::new(sys::TunSocket::new(name).expect("TunSocket::new")),
        }
    }
    pub fn name(&self) -> io::Result<String> {
        self.watcher.get_ref().name()
    }

    pub fn mtu(&self) -> io::Result<usize> {
        self.watcher.get_ref().mtu()
    }
}

impl AsyncRead for TunSocket {
    fn poll_read(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &mut [u8],
    ) -> Poll<io::Result<usize>> {
        Pin::new(&mut &*self).poll_read(cx, buf)
    }
}

impl AsyncRead for &TunSocket {
    fn poll_read(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &mut [u8],
    ) -> Poll<io::Result<usize>> {
        self.watcher.poll_read_with(cx, |mut inner| inner.read(buf))
    }
}

impl AsyncWrite for TunSocket {
    fn poll_write(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &[u8],
    ) -> Poll<io::Result<usize>> {
        Pin::new(&mut &*self).poll_write(cx, buf)
    }

    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
        Pin::new(&mut &*self).poll_flush(cx)
    }

    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
        Pin::new(&mut &*self).poll_close(cx)
    }
}

impl AsyncWrite for &TunSocket {
    fn poll_write(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &[u8],
    ) -> Poll<io::Result<usize>> {
        self.watcher
            .poll_write_with(cx, |mut inner| inner.write(buf))
    }

    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
        self.watcher.poll_write_with(cx, |mut inner| inner.flush())
    }

    fn poll_close(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<io::Result<()>> {
        Poll::Ready(Ok(()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_std::net::UdpSocket;
    use async_std::task::block_on;
    use futures::AsyncReadExt;
    use insta::assert_debug_snapshot;
    use sysconfig::setup_ip;

    #[test]
    fn test_recv_packets_from_tun() {
        let tun_name = "utun3";
        let mut tun_socket = TunSocket::new(tun_name);
        setup_ip(tun_name, "10.0.1.1", "10.0.1.0/24");

        let data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 19];

        block_on(async move {
            let socket = UdpSocket::bind("0.0.0.0:0").await.unwrap();
            socket.send_to(&data, "10.0.1.2:80").await.unwrap();
            let mut buf = vec![0; 1024];
            let size = tun_socket.read(&mut buf).await.unwrap();

            assert_debug_snapshot!(&buf[(size - data.len())..size]);
        })
    }
}
