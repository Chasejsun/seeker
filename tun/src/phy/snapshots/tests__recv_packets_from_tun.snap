---
source: tun/src/phy/mod.rs
expression: "&buf[(size - data.len())..size]"
---
[
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    19,
]
